packer/provisioner.go

```diff
@@ -145,15 +145,70 @@ func (p *PausedProvisioner) Prepare(raws ...interface{}) error {
 
 func (p *PausedProvisioner) Provision(ctx context.Context, ui packersdk.Ui, comm packersdk.Communicator, generatedData map[string]interface{}) error {
 
-	// Use a select to determine if we get cancelled during the wait
 	ui.Say(fmt.Sprintf("Pausing %s before the next provisioner...", p.PauseBefore))
+
+	var err error
+	switch ui.(type) {
+	case *MachineReadableUi:
+		err = p.pausingNoUpdate(ctx)
+	default:
+		minimumTimeForPauseUpdate := float64(10)
+		if p.PauseBefore.Seconds() < minimumTimeForPauseUpdate {
+			err = p.pausingNoUpdate(ctx)
+		} else {
+			err = p.pausingWithUpdates(ctx, ui)
+		}
+	}
+	if err != nil {
+		return err
+	}
+
+	return p.Provisioner.Provision(ctx, ui, comm, generatedData)
+}
+
+func (p *PausedProvisioner) pausingNoUpdate(ctx context.Context) error {
 	select {
 	case <-time.After(p.PauseBefore):
 	case <-ctx.Done():
 		return ctx.Err()
 	}
+	return nil
+}
+
+func (p *PausedProvisioner) pausingWithUpdates(ctx context.Context, ui packersdk.Ui) error {
+	updateTime := 10
+	var C <-chan time.Time
+	if updateTime > 0 {
+		ticker := time.NewTicker(time.Duration(updateTime) * time.Second)
+		defer ticker.Stop()
+	}
+	totalTime := p.PauseBefore.Seconds()
+
+	pausingCtx, cancel := context.WithTimeout(ctx, p.PauseBefore)
+	defer cancel()
+
+	var err error
+	for {
+		select {
+		case <-pausingCtx.Done():
+			err = pausingCtx.Err()
+			if err != nil && isTimeRemaining(totalTime) {
+				return err
+			}
+			return nil
+		case <-ticker.C:
```

Comment:
To help with the fix you added:
```suggestion
		case <- C:
```

Reply:
I implemented this but the `case<-C` is never being hit in the select statement, is there something I'm missing? Also is there a reason we don't want to use the ticker, I'm confused why we shouldn't use that in the select statement.

Reply:
My suggestion was to make the code similar for both cases, if there is no timeout then the chan is nil and never unblocks, ( which is the current behaviour ) otherwise it prints stuff; that'd make the PR much more simple.

Reply:
@azr I made the changes, please let me know if there is anything else I should change.